mcos/optimizer.py

```diff
@@ -253,30 +253,40 @@ def _cluster_var(self, cov: np.ndarray) -> np.ndarray:
         w_ = self._inverse_variance_weights(cov).reshape(-1, 1)
         return np.dot(np.dot(w_.T, cov), w_)[0, 0]
 
-    def _hrp_weights(self, cov: np.ndarray, sorted_indices: List) -> np.ndarray:
-        """
-        Gets position weights using hierarchical risk parity
-        :param cov: covariance matrix
-        :param sorted_indices: clustering scheme
-        :return: array of position weights
+    def _hrp_weights(self,cov: np.ndarray, sorted_indices: List) -> np.ndarray:
         """
+      Gets position weights using hierarchical risk parity
+      :param cov: covariance matrix
+      :param sorted_indices: clustering scheme
+      :return: array of position weights, sorted to match the original order passed
+      """
+
         if len(sorted_indices) == 0:
             raise ValueError('sorted_indices is empty')
 
         if len(sorted_indices) == 1:
             return np.array([1.])
 
-        split_indices = np.array_split(np.array(sorted_indices), 2)
+        weights = pd.Series(1, index=sorted_indices)
+        clusters = [sorted_indices]
+
+        while len(clusters) > 0:
```

Comment:
Could you explain why this produces different results than `np.array_split`?

Reply:
If there are an odd number of securities, array_split favors the left branch, whereas this code favors the right. So splitting [1,2,3,4,5] with array_split gives [1,2,3],[4,5] where this code gives [1,2],[3,4,5]. Sorry if I missed something that compensates for that, I just wanted to make sure we were following the paper since that split has a big effect on the weights.